File: apps/ios/notiving/notiving/Shell/Bridge/RN/ShellBridgeModule.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(ShellBridge, NSObject)

// Session
RCT_EXTERN_METHOD(getToken:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getUserId:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(addListener:(NSString *)eventName)
RCT_EXTERN_METHOD(removeListeners:(NSInteger)count)

// Navigation
RCT_EXTERN_METHOD(navigate:(NSString *)url opts:(NSDictionary *)opts)
RCT_EXTERN_METHOD(back)
RCT_EXTERN_METHOD(setTitle:(NSString *)title)
RCT_EXTERN_METHOD(setNavBarHidden:(BOOL)hidden)

// Analytics
RCT_EXTERN_METHOD(track:(NSString *)event params:(NSDictionary *)params)
RCT_EXTERN_METHOD(pageView:(NSString *)pageName)

// Permission
RCT_EXTERN_METHOD(requestPermission:(NSString *)type resolve:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(checkPermission:(NSString *)type resolve:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

// Lifecycle
RCT_EXTERN_METHOD(ready)

// Device
RCT_EXTERN_METHOD(getDeviceInfo:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getAppVersion:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end

File: apps/ios/notiving/notiving/Shell/Bridge/RN/ShellBridgeModule.swift
import Foundation

@objc(ShellBridge)
final class ShellBridgeModule: NSObject {

    @objc static func requiresMainQueueSetup() -> Bool { false }

    @objc static func supportedEvents() -> [String] {
        return ["onSessionChange", "onResume", "onPause"]
    }

    // --- Session ---

    @objc func getToken(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        resolve(SessionManager.shared.accessToken)
    }

    @objc func getUserId(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        resolve(SessionManager.shared.userId)
    }

    @objc func addListener(_ eventName: String) {
        // Required for RN EventEmitter
    }

    @objc func removeListeners(_ count: Int) {
        // Required for RN EventEmitter
    }

    // --- Navigation ---

    @objc func navigate(_ url: String, opts: NSDictionary?) {
        DispatchQueue.main.async {
            ShellRouter.shared.navigate(to: url)
        }
    }

    @objc func back() {
        DispatchQueue.main.async {
            ShellRouter.shared.back()
        }
    }

    @objc func setTitle(_ title: String) {
        DispatchQueue.main.async {
            ShellRouter.shared.setTitle(title)
        }
    }

    @objc func setNavBarHidden(_ hidden: Bool) {
        DispatchQueue.main.async {
            ShellRouter.shared.setNavBarHidden(hidden)
        }
    }

    // --- Analytics ---

    @objc func track(_ event: String, params: NSDictionary?) {
        // Phase 4: forward to AnalyticsTracker
    }

    @objc func pageView(_ pageName: String) {
        // Phase 4: forward to AnalyticsTracker
    }

    // --- Permission ---

    @objc func requestPermission(
        _ type: String,
        resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        // Phase 4: forward to PermissionManager
        resolve(["status": "denied"])
    }

    @objc func checkPermission(
        _ type: String,
        resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        // Phase 4: forward to PermissionManager
        resolve(["status": "denied"])
    }

    // --- Lifecycle ---

    @objc func ready() {
        // Container signals first render complete. Phase 4: performance tracking.
    }

    // --- Device ---

    @objc func getDeviceInfo(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        let info: [String: String] = [
            "platform": "ios",
            "osVersion": UIDevice.current.systemVersion,
            "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "",
            "deviceId": SessionManager.shared.deviceId ?? "",
        ]
        resolve(info)
    }

    @objc func getAppVersion(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
        resolve(version)
    }
}
